pkg/cloudlogging/puller.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package cloudlogging pulls lumberjack/data access logs from GCP cloud
+// logging.
+package cloudlogging
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"cloud.google.com/go/logging/apiv2/loggingpb"
+	"github.com/sethvargo/go-retry"
+	"google.golang.org/api/iterator"
+
+	logging "cloud.google.com/go/logging/apiv2"
+)
+
+// Puller pulls log entries of GCP organizations, folders, projects, or
+// billingAccounts.
+type Puller struct {
+	client *logging.Client
+	// Optional retry backoff strategy, default is 5 attempts with fibonacci
+	// backoff that starts at 500ms.
+	retry retry.Backoff
+}
+
+// Option is the option to set up an log puller.
+type Option func(h *Puller) *Puller
+
+// WithRetry provides retry strategy to the log puller.
+func WithRetry(b retry.Backoff) Option {
+	return func(p *Puller) *Puller {
+		p.retry = b
+		return p
+	}
+}
+
+// NewPuller creates a new Puller with provided clients and options.
+func NewPuller(ctx context.Context, c *logging.Client, opts ...Option) *Puller {
+	p := &Puller{client: c}
+	for _, opt := range opts {
+		p = opt(p)
+	}
+
+	if p.retry == nil {
+		p.retry = retry.WithMaxRetries(5, retry.NewFibonacci(500*time.Millisecond))
+	}
+	return p
+}
+
+// Pull pulls a list of log entries given request.
+func (p *Puller) Pull(ctx context.Context, req *loggingpb.ListLogEntriesRequest) ([]*loggingpb.LogEntry, error) {
```

Comment:
I wonder if we should have a `DefaultPull` with a preset `req`. The `req` will be based on log name.

Reply:
Why? I am thinking this can live in the cli command?

Reply:
I'm fine with that as well

Reply:
Actually I changed my mind - I think it's better if we make `filter` an input here. A `Puller` should be scoped to a project/folder/org on initialization.

The `Pull` here is essential "pull once and see what you got" operation. This is what we immediately need. I'm thinking maybe later we can have a better approach to "tail the logs and see if any matching logs coming up".

Reply:
I am not sure I follow, I am sharing my [plan](https://github.com/abcxyz/lumberjack/blob/sqin/cli/pkg/cli/pull_validate.go#L191) here - we can form the request in the cli however we want. For scoping the Puller, there is only one cloud logging client though. We can have a quick sync if you want.

Reply:
ah..I got that scoping the Puller part, but I am wondering why we should do that...